File: src/engines/__tests__/shortDivision.test.ts
import { describe, it, expect } from 'vitest';
import { generateShortDivisionSteps } from '../shortDivision';

describe('generateShortDivisionSteps', () => {
  it('handles exact division with no remainders', () => {
    // 84 ÷ 7 = 12
    const steps = generateShortDivisionSteps(7, 84);
    const answerSteps = steps.filter(s => s.type === 'answer_digit');
    const remainderSteps = steps.filter(s => s.type === 'remainder');

    expect(answerSteps).toHaveLength(2);
    expect(answerSteps[0].correctValue).toBe(1); // 8÷7 = 1
    expect(answerSteps[1].correctValue).toBe(2); // 14÷7 = 2
    expect(remainderSteps).toHaveLength(1); // carry remainder from first digit
    expect(remainderSteps[0].correctValue).toBe(1); // 8%7 = 1
  });

  it('handles 98 ÷ 7 = 14', () => {
    const steps = generateShortDivisionSteps(7, 98);
    const answerSteps = steps.filter(s => s.type === 'answer_digit');

    expect(answerSteps[0].correctValue).toBe(1); // 9÷7 = 1
    expect(answerSteps[1].correctValue).toBe(4); // 28÷7 = 4
  });

  it('handles division with final remainder', () => {
    // 85 ÷ 7 = 12 r1
    const steps = generateShortDivisionSteps(7, 85);
    const answerSteps = steps.filter(s => s.type === 'answer_digit');
    const remainderSteps = steps.filter(s => s.type === 'remainder');

    expect(answerSteps[0].correctValue).toBe(1); // 8÷7 = 1
    expect(answerSteps[1].correctValue).toBe(2); // 15÷7 = 2
    // One carry remainder (from 8%7=1) and one final remainder (15%7=1)
    const finalRemainder = remainderSteps.find(s => s.label === 'r');
    expect(finalRemainder).toBeDefined();
    expect(finalRemainder!.correctValue).toBe(1);
  });

  it('handles 3-digit dividend', () => {
    // 532 ÷ 4 = 133
    const steps = generateShortDivisionSteps(4, 532);
    const answerSteps = steps.filter(s => s.type === 'answer_digit');

    expect(answerSteps[0].correctValue).toBe(1); // 5÷4 = 1
    expect(answerSteps[1].correctValue).toBe(3); // 13÷4 = 3
    expect(answerSteps[2].correctValue).toBe(3); // 12÷4 = 3
  });

  it('handles quotient digit of 0', () => {
    // 203 ÷ 4 = 50 r3
    const steps = generateShortDivisionSteps(4, 203);
    const answerSteps = steps.filter(s => s.type === 'answer_digit');

    expect(answerSteps[0].correctValue).toBe(0); // 2÷4 = 0
    expect(answerSteps[1].correctValue).toBe(5); // 20÷4 = 5
    expect(answerSteps[2].correctValue).toBe(0); // 3÷4 = 0
  });
});

File: go-engines/short_division_test.go
package engines

import "testing"

func TestGenerateShortDivisionSteps_ExactDivision(t *testing.T) {
	// 84 ÷ 7 = 12
	steps := GenerateShortDivisionSteps(7, 84)
	answerSteps := filterStepsByType(steps, AnswerDigit)
	remainderSteps := filterStepsByType(steps, RemainderStep)

	if len(answerSteps) != 2 {
		t.Fatalf("expected 2 answer steps, got %d", len(answerSteps))
	}
	if answerSteps[0].CorrectValue != 1 {
		t.Errorf("first quotient digit: expected 1, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 2 {
		t.Errorf("second quotient digit: expected 2, got %d", answerSteps[1].CorrectValue)
	}
	if len(remainderSteps) != 1 {
		t.Fatalf("expected 1 remainder step, got %d", len(remainderSteps))
	}
	if remainderSteps[0].CorrectValue != 1 {
		t.Errorf("remainder carry: expected 1, got %d", remainderSteps[0].CorrectValue)
	}
}

func TestGenerateShortDivisionSteps_98Div7(t *testing.T) {
	steps := GenerateShortDivisionSteps(7, 98)
	answerSteps := filterStepsByType(steps, AnswerDigit)

	if answerSteps[0].CorrectValue != 1 {
		t.Errorf("first: expected 1, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 4 {
		t.Errorf("second: expected 4, got %d", answerSteps[1].CorrectValue)
	}
}

func TestGenerateShortDivisionSteps_WithFinalRemainder(t *testing.T) {
	// 85 ÷ 7 = 12 r1
	steps := GenerateShortDivisionSteps(7, 85)
	answerSteps := filterStepsByType(steps, AnswerDigit)
	remainderSteps := filterStepsByType(steps, RemainderStep)

	if answerSteps[0].CorrectValue != 1 {
		t.Errorf("first: expected 1, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 2 {
		t.Errorf("second: expected 2, got %d", answerSteps[1].CorrectValue)
	}

	var finalRemainder *StepInput
	for i := range remainderSteps {
		if remainderSteps[i].Label == "r" {
			finalRemainder = &remainderSteps[i]
			break
		}
	}
	if finalRemainder == nil {
		t.Fatal("expected a final remainder step with label 'r'")
	}
	if finalRemainder.CorrectValue != 1 {
		t.Errorf("final remainder: expected 1, got %d", finalRemainder.CorrectValue)
	}
}

func TestGenerateShortDivisionSteps_3DigitDividend(t *testing.T) {
	// 532 ÷ 4 = 133
	steps := GenerateShortDivisionSteps(4, 532)
	answerSteps := filterStepsByType(steps, AnswerDigit)

	if answerSteps[0].CorrectValue != 1 {
		t.Errorf("first: expected 1, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 3 {
		t.Errorf("second: expected 3, got %d", answerSteps[1].CorrectValue)
	}
	if answerSteps[2].CorrectValue != 3 {
		t.Errorf("third: expected 3, got %d", answerSteps[2].CorrectValue)
	}
}

func TestGenerateShortDivisionSteps_QuotientDigitZero(t *testing.T) {
	// 203 ÷ 4 = 50 r3
	steps := GenerateShortDivisionSteps(4, 203)
	answerSteps := filterStepsByType(steps, AnswerDigit)

	if answerSteps[0].CorrectValue != 0 {
		t.Errorf("first: expected 0, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 5 {
		t.Errorf("second: expected 5, got %d", answerSteps[1].CorrectValue)
	}
	if answerSteps[2].CorrectValue != 0 {
		t.Errorf("third: expected 0, got %d", answerSteps[2].CorrectValue)
	}
}
